pilot/pkg/serviceregistry/kube/controller/discoverycontrollers.go

```diff
@@ -0,0 +1,220 @@
+	kubeClient kubelib.Client,
+	endpointMode EndpointMode,
+	meshWatcher mesh.Watcher,
+	discoveryNamespacesFilter filter.DiscoveryNamespacesFilter,
+) {
+	meshWatcher.AddMeshHandler(func() {
+		newSelectedNamespaces, deselectedNamespaces := discoveryNamespacesFilter.SelectorsChanged(meshWatcher.Mesh().GetDiscoverySelectors())
+
+		for _, nsName := range newSelectedNamespaces {
+			c.queue.Push(func() error {
+				c.handleSelectedNamespace(endpointMode, nsName)
+				return nil
+			})
+		}
+
+		for _, nsName := range deselectedNamespaces {
+			c.queue.Push(func() error {
+				c.handleDeselectedNamespace(kubeClient, endpointMode, nsName)
+				return nil
+			})
+		}
+	})
+}
+
+// issue create events for all services, pods, and endpoints in the newly labeled namespace
+func (c *Controller) handleSelectedNamespace(endpointMode EndpointMode, ns string) {
+	var errs *multierror.Error
+
+	// for each resource type, issue create events for objects in the labeled namespace
+
+	services, err := c.serviceLister.Services(ns).List(labels.Everything())
+	if err != nil {
+		log.Errorf("error listing services: %v", err)
+		return
+	}
+	for _, svc := range services {
+		errs = multierror.Append(errs, c.onServiceEvent(svc, model.EventAdd))
+	}
+
+	pods, err := listerv1.NewPodLister(c.pods.informer.GetIndexer()).Pods(ns).List(labels.Everything())
+	if err != nil {
+		log.Errorf("error listing pods: %v", err)
+		return
+	}
+	for _, pod := range pods {
+		errs = multierror.Append(errs, c.pods.onEvent(pod, model.EventAdd))
+	}
+
+	switch endpointMode {
+	case EndpointsOnly:
+		endpoints, err := listerv1.NewEndpointsLister(c.endpoints.getInformer().GetIndexer()).Endpoints(ns).List(labels.Everything())
+		if err != nil {
+			log.Errorf("error listing endpoints: %v", err)
+			return
+		}
+		for _, ep := range endpoints {
+			errs = multierror.Append(errs, c.endpoints.onEvent(ep, model.EventAdd))
+		}
+	case EndpointSliceOnly:
+		endpointSlices, err := v1beta1.NewEndpointSliceLister(c.endpoints.getInformer().GetIndexer()).EndpointSlices(ns).List(labels.Everything())
+		if err != nil {
+			log.Errorf("error listing endpoint slices: %v", err)
+			return
+		}
+		for _, ep := range endpointSlices {
+			errs = multierror.Append(errs, c.endpoints.onEvent(ep, model.EventAdd))
+		}
+	}
+
+	if err := multierror.Flatten(errs.ErrorOrNil()); err != nil {
+		log.Errorf("one or more errors while handling newly labeled discovery namespace %s: %v", ns, err)
+	}
+}
+
+// issue delete events for all services, pods, and endpoints in the delabled namespace
+// use kubeClient to bypass filter in order to list resources from non-labeled namespace
+func (c *Controller) handleDeselectedNamespace(kubeClient kubelib.Client, endpointMode EndpointMode, ns string) {
+	var errs *multierror.Error
+
+	// for each resource type, issue delete events for objects in the delabled namespace
+
+	services, err := kubeClient.KubeInformer().Core().V1().Services().Lister().Services(ns).List(labels.Everything())
```

Comment:
nit: we might potentially want a GetUnfiltered() method on the filter interface, so we don't have to create a new informer here? I realize its cached under the hood but it still "looks" wrong at first and is pretty subtle. If not, a comment here indicating its cached would help

Reply:
i'll add a comment reminding that it's cached and not instantiating a new informer